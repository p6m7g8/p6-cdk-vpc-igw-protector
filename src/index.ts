export * from './p6cdkvpcigwprotector'
