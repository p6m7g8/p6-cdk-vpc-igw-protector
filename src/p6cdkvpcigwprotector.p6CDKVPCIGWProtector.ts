import type { DetachInternetGatewayCommandInput } from '@aws-sdk/client-ec2'
import type { Context } from 'aws-lambda'
import type { Logger } from 'winston'
import { EC2 } from '@aws-sdk/client-ec2'
import winston from 'winston'

// Configure the logger
const logger: Logger = winston.createLogger({
  level: 'info',
  format: winston.format.json(),
  defaultMeta: { service: 'VPC-remediation' },
  transports: [new winston.transports.Console()],
})

const ec2 = new EC2({})

// Define the structure of the event details
interface VPCEvent {
  detail: {
    requestParameters: {
      internetGatewayId: DetachInternetGatewayCommandInput['InternetGatewayId']
      vpcId: DetachInternetGatewayCommandInput['VpcId']
    }
    errorCode?: string
    errorMessage?: string
  }
}

// Loop prevention function
function p6LoopPrevent(event: VPCEvent): boolean {
  if (event.detail.errorCode || event.detail.errorMessage) {
    logger.info('Previous API call resulted in an error. Ending')
    return true
  }
  return false
}

// Detach and delete Internet Gateway from VPC if not authorized
async function p6Ec2VpcInternetGatewayAuthorizedVpcOnly(event: VPCEvent): Promise<void> {
  if (p6LoopPrevent(event))
    return

  const igwId = event.detail.requestParameters.internetGatewayId
  const vpcId = event.detail.requestParameters.vpcId

  logger.info(`Processing Internet Gateway: ${igwId}`)
  try {
    // Detach IGW from VPC
    await ec2.detachInternetGateway({ InternetGatewayId: igwId, VpcId: vpcId })
    logger.info(`Detached IGW ${igwId} from VPC ${vpcId}`)

    // Delete IGW
    await ec2.deleteInternetGateway({ InternetGatewayId: igwId })
    logger.info(`Deleted IGW ${igwId}`)
  }
  catch (error) {
    logger.error(`Failed to process IGW ${igwId}: ${error}`)
  }
}

// Lambda handler
export async function handler(event: VPCEvent, _context?: Context): Promise<boolean> {
  await p6Ec2VpcInternetGatewayAuthorizedVpcOnly(event)
  return true
}

// Main function for developer testing
export async function main(): Promise<void> {
  logger.debug('Testing handler()')
  await handler({
    detail: {
      requestParameters: {
        internetGatewayId: 'igw-1234567890',
        vpcId: 'vpc-1234567890',
      },
    },
  })
}

// Execute main function if script is run directly
if (require.main === module) {
  main()
}
