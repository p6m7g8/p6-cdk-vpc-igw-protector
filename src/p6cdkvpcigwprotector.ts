import type { Construct } from 'constructs'
import * as cdk from 'aws-cdk-lib'
import * as events from 'aws-cdk-lib/aws-events'
import * as targets from 'aws-cdk-lib/aws-events-targets'
import * as lambda from 'aws-cdk-lib/aws-lambda'
import * as lambdajs from 'aws-cdk-lib/aws-lambda-nodejs'
import * as floyd from 'cdk-iam-floyd'

export interface IP6CDKVPCIGWProtectorProps {
}

export class P6CDKVPCIGWProtector extends cdk.Resource {
  constructor(scope: Construct, id: string, _props: IP6CDKVPCIGWProtectorProps) {
    super(scope, id)

    const policy = new floyd.Statement.Ec2()
      .allow()
      .toDetachInternetGateway()
      .toDeleteInternetGateway()
      .toDescribeVpcs()
      .toDescribeInternetGateways()
      .on('*')

    const fn = new lambdajs.NodejsFunction(this, 'p6CDKVPCIGWProtector', {
      runtime: lambda.Runtime.NODEJS_20_X,
      timeout: cdk.Duration.seconds(5),
      tracing: lambda.Tracing.ACTIVE,
      bundling: {
        minify: true,
        externalModules: ['aws-sdk'],
      },
    })

    fn.addToRolePolicy(policy)

    const rule = new events.Rule(this, 'vpcInternetGatewayAuthorizedVpcOnlyRule', {
      description: 'Detect unexpected Vpc Internet Gateways and remediates',
      eventPattern: {
        source: ['aws.ec2'],
        detail: {
          eventSource: ['ec2.amazonaws.com'],
          eventName: ['AttachInternetGateway'],
        },
      },
    })

    rule.addTarget(new targets.LambdaFunction(fn))
  }
}
