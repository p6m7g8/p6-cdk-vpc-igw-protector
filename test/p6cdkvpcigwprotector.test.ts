import * as cdk from 'aws-cdk-lib'
import { Template } from 'aws-cdk-lib/assertions'
import { P6CDKVPCIGWProtector } from '../src'

it('p6CDKVPCIGWProtector snapshot test', () => {
  const app = new cdk.App()
  const stack = new cdk.Stack(app, 'MyStack')

  // WHEN
  new P6CDKVPCIGWProtector(stack, 'p6-cdk-vpc-igw-protector', {})

  // THEN
  const template = Template.fromStack(stack).toJSON()
  expect(template).toMatchSnapshot()
})
