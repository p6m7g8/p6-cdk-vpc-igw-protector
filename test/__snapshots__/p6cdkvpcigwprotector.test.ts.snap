// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6CDKVPCIGWProtector snapshot test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "p6cdkvpcigwprotectorp6CDKVPCIGWProtector23CE1718": {
      "DependsOn": [
        "p6cdkvpcigwprotectorp6CDKVPCIGWProtectorServiceRoleDefaultPolicyB91EFBAD",
        "p6cdkvpcigwprotectorp6CDKVPCIGWProtectorServiceRole79E45E64",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "3c388371783f09cc298c0b4d2191de48c7ab204b29dc446f07becf8da02f5f9a.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "p6cdkvpcigwprotectorp6CDKVPCIGWProtectorServiceRole79E45E64",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 5,
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "p6cdkvpcigwprotectorp6CDKVPCIGWProtectorServiceRole79E45E64": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "p6cdkvpcigwprotectorp6CDKVPCIGWProtectorServiceRoleDefaultPolicyB91EFBAD": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "ec2:DeleteInternetGateway",
                "ec2:DescribeInternetGateways",
                "ec2:DescribeVpcs",
                "ec2:DetachInternetGateway",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "p6cdkvpcigwprotectorp6CDKVPCIGWProtectorServiceRoleDefaultPolicyB91EFBAD",
        "Roles": [
          {
            "Ref": "p6cdkvpcigwprotectorp6CDKVPCIGWProtectorServiceRole79E45E64",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "p6cdkvpcigwprotectorvpcInternetGatewayAuthorizedVpcOnlyRule5FE5A2BB": {
      "Properties": {
        "Description": "Detect unexpected Vpc Internet Gateways and remediates",
        "EventPattern": {
          "detail": {
            "eventName": [
              "AttachInternetGateway",
            ],
            "eventSource": [
              "ec2.amazonaws.com",
            ],
          },
          "source": [
            "aws.ec2",
          ],
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "p6cdkvpcigwprotectorp6CDKVPCIGWProtector23CE1718",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "p6cdkvpcigwprotectorvpcInternetGatewayAuthorizedVpcOnlyRuleAllowEventRuleMyStackp6cdkvpcigwprotectorp6CDKVPCIGWProtectorE270487FC72DF2A9": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "p6cdkvpcigwprotectorp6CDKVPCIGWProtector23CE1718",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "p6cdkvpcigwprotectorvpcInternetGatewayAuthorizedVpcOnlyRule5FE5A2BB",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
